File: Pragma_Examples/OpenMP/Fortran/8_jacobi/1_jacobi_usm/norm.f90
module norm_mod
  use kind_mod
  use mesh_mod, only: mesh_t
  implicit none
  !$omp requires unified_shared_memory

  private
  public :: norm

contains

  function norm(mesh, u) result(norm_val)
    type(mesh_t), intent(inout) :: mesh
    real(RK), intent(inout) :: u(:,:)
    real(RK) :: norm_val
    integer(IK) :: i,j,n_x,n_y
    real(RK) :: dxdy

    dxdy = mesh%dx*mesh%dy

    norm_val = 0._RK

    !$omp target teams distribute parallel do collapse(2) reduction(+:norm_val)
    do j = 1,mesh%n_y
      do i = 1,mesh%n_x
        norm_val = norm_val + u(i,j)**2*dxdy
      end do
    end do

    norm_val = sqrt(norm_val)/(mesh%n_x*mesh%n_y)
  end function norm

end module norm_mod

File: Pragma_Examples/OpenMP/CXX/8_jacobi/1_jacobi_usm/Norm.cpp
//**************************************************************************
//* Copyright (c) 2023, Advanced Micro Devices, Inc. All rights reserved.
//**************************************************************************

#include "Jacobi.hpp"

#pragma omp requires unified_shared_memory
dfloat Norm(mesh_t& mesh, dfloat *U) 
{
  dfloat norm = 0.0;
  const int N = mesh.N;
  const dfloat dx = mesh.dx;
  const dfloat dy = mesh.dy;
  #pragma omp target teams distribute parallel for reduction(+:norm)
  for (int id=0; id < N; id++) {
    norm += U[id] * U[id] * dx * dy;
  }
  return sqrt(norm)/N;
}
